### PR TITLE
Integrate Foundation NFT Marketplace

### PR DESCRIPTION
**Pull request by lucasespinosa28**
I did this plugin for the hackathon, but I still needed to fix some UI issues, is now working.

Screenshot from 2021-10-03 13-10-17
https://gitcoin.co/issue/DimensionDev/Maskbook/4298/100026557

### CODE REVIEW

File: packages/maskbook/src/plugins/Foudation/utils.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+export function convertDate(unix: string) {
+    const date = new Date(Number(unix) * 1000)
+    const humanDateFormat = date.toLocaleString()
+    return humanDateFormat
+}
+
+export function shortAddress(address: string) {
+    return `${address.slice(0, 6)}...${address.slice(-4)}`
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/utils.ts:59]: Please search the `formatAddress` in the codebase.

File: packages/maskbook/src/plugins/Foudation/utils.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    if (address) {
+        const pathSlit = address.split('/')
+        const pathTwo = pathSlit[pathSlit.length - 2].slice(-4)
+        return [pathTwo.slice(0, 2), pathTwo.slice(2), pathSlit[pathSlit.length - 2]]
+    }
+    return ['']
+}
+
+export function convertDate(unix: string) {
+    const date = new Date(Number(unix) * 1000)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/utils.ts:53]:
```suggestion
    const date = new Date(Number.parseInt(unix, 10) * 1000)
```

File: packages/maskbook/src/plugins/Foudation/types/index.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    id: string
+}
+
+export interface NftHistory {
+    amountInETH: string
+    contractAddress: string
+    date: string
+    event: string
+    id: string
+    tokenAddress?: any
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/types/index.ts:43]: Avoid `any` if possible.

File: packages/maskbook/src/plugins/Foudation/abis/Foundation.json [hunk 4]
```diff
@@ -0,0 +1,733 @@
+[
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/abis/Foundation.json:1]: Move this file to `web3-contracts/abis`

File: packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoundationPlaceBid.tsx [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    const hasnHighestBid = () => {
+        if (props.nftData.graph.data.nfts[0].mostRecentAuction.highestBid) {
+            return props.nftData.graph.data.nfts[0].mostRecentAuction.highestBid.bidder.id.includes(account)
+        }
+        return false
+    }
+    const [amount, setAmount] = useState<string>('')
+    const contractAddres =
+        props.nftData.chainId === ChainId.Gorli
+            ? '0xeB1bD095061bbDb1aD065524628812cae63e4222'
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoundationPlaceBid.tsx:51]: Move to `web3-constants`.

File: packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationProvenances.tsx [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        if (a.date > b.date) {
+            return -1
+        }
+        return 0
+    })
+    return (
+        <Paper>
+            {props.histories.map((history: NftHistory) => (
+                <>
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationProvenances.tsx:27]: The `key`  is lost.

File: packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationProvenances.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+interface Props extends React.PropsWithChildren<{}> {
+    histories: NftHistory[]
+}
+
+function FoudationProvenances(props: Props) {
+    const { t } = useI18N()
+    const account = useAccount()
+    props.histories.sort((a: NftHistory, b: NftHistory) => {
+        if (a.date < b.date) {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationProvenances.tsx:16]: `return b.date - a.date`

File: packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationDescription.tsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { Typography } from '@material-ui/core'
+
+interface Props extends React.PropsWithChildren<{}> {
+    description: string
+}
+
+function FoudationDescription(props: Props) {
+    return (
+        <>
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationDescription.tsx:9]: Remove the fragment element.

File: packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationCountdown.tsx [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        setCount({
+            hours: Math.floor((distance % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60)),
+            minutes: Math.floor((distance % (1000 * 60 * 60)) / (1000 * 60)),
+            seconds: Math.floor((distance % (1000 * 60)) / 1000),
+        })
+    }
+    const NftdateEnding = new Date(Number(props.auctions.at(-1)?.dateEnding) * 1000).getTime()
+    useEffect(() => {
+        if (NftdateEnding > Date.now()) {
+            setInterval(timer, 1000)
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationCountdown.tsx:42]: Need to clean the timer.

File: packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationCard.tsx [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            height: 10,
+        },
+    }
+})
+
+function FoudationCard(props: Props) {
+    const { classes } = useStyles()
+    const { t } = useI18N()
+    const [nftData, setNftData] = useState<nftData>()
+    useEffect(() => {
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/SNSAdaptor/FoudationCard.tsx:64]: Try to use `useAsync`  or `useAsyncRetry` from `react-use`.

File: packages/maskbook/src/plugins/Foudation/Components.tsx [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import { Box } from '@material-ui/core'
```
Comment (guanbinrui) [inline packages/maskbook/src/plugins/Foudation/Components.tsx:1]: Remove this file if it's not necessary.

**Review by lucasespinosa28 — COMMENTED** [review]
added recommended code changes

**Review by lucasespinosa28 — COMMENTED** [review]
I add the callback and the fetch in the background

the UI was improved

Screenshot from 2021-10-11 02-44-22
Screenshot from 2021-10-11 02-52-00
Screenshot from 2021-10-11 02-54-08
Screenshot from 2021-10-11 03-29-16
Screenshot from 2021-10-11 03-58-31
Screenshot from 2021-10-11 03-59-00

**Review by lucasespinosa28 — COMMENTED** [review]
review

**Review by lucasespinosa28 — COMMENTED** [review]
problem loading images , solved

**Review by lucasespinosa28 — COMMENTED** [review]
misspelled name

**Review by Jack-Works — CHANGES_REQUESTED** [review]
I noticed you have changed some file permissions from 100644 to 100755. Please revert that, thanks

**Review by lucasespinosa28 — COMMENTED** [review]
the issue with files changed to 100644 to 100755 has been fixed

**Review by lucasespinosa28 — COMMENTED** [review]
I fixed the countdown and the error with texts

**Review by lucasespinosa28 — COMMENTED** [review]
refactor countdown

**Review by lucasespinosa28 — COMMENTED** [review]
I fixed the conflicts

**Review by lucasespinosa28 — COMMENTED** [review]
sorry for the delay,i fixed the big conflict

**Review by lucasespinosa28 — COMMENTED** [review]
fix conflict

**Review by lucasespinosa28 — COMMENTED** [review]
I fixed the plugin path

File: packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationCountdown.tsx [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+            const timer = setInterval(() => {
+                const now = Date.now()
+                const distance = nftDateEnding - now
+                setCount({
+                    days: Math.floor(distance / (1000 * 60 * 60 * 24)),
+                    hours: Math.floor((distance % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60)),
+                    minutes: Math.floor((distance % (1000 * 60 * 60)) / (1000 * 60)),
+                    seconds: Math.floor((distance % (1000 * 60)) / 1000),
+                })
+                console.log(currentCount)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationCountdown.tsx:45]:
```suggestion
```

File: packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationCountdown.tsx [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+            }, 1000)
+            return () => {
+                clearTimeout(timer)
+            }
+        }
+        setEnded(true)
+        return
+    }, [currentCount])
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationCountdown.tsx:53]:
```suggestion
    }, [currentCount, nftDateEnding])
```

File: packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationHeader.tsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+    const CurrentPrice = useCallback(() => {
+        if (props.nft.mostRecentAuction) {
+            if (props.nft.mostRecentAuction.highestBid) {
+                return `${t('plugin_foundation_highest')} ${props.nft.mostRecentAuction.highestBid.amountInETH}`
+            }
+            return `${t('plugin_foundation_reserve')} ${props.nft.mostRecentAuction.reservePriceInETH}`
+        }
+        return null
+    }, [])
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationHeader.tsx:23]:
```suggestion
    }, [props.nft.mostRecentAuction])
```

File: packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationProvenances.tsx [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+    const account = useAccount()
+    props.histories.sort((first: NftHistory, second: NftHistory) => {
+        if (first.date < second.date) {
+            return 1
+        }
+        if (first.date > second.date) {
+            return -1
+        }
+        return 0
+    })
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationProvenances.tsx:25]: Sorting an array is a side-effect. Therefore, please create a duplication before sorting it.

File: packages/mask/src/plugins/Foundation/hooks/usePlaceBidCallback.ts [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+        if (!MARKET_ADDRESS) return []
+        const bidAmount = amount
+        const bidAuctionId = auctionId
+        return [
+            bidAuctionId, // id of nft in the market
+            bidAmount, // bid price
+        ]
+    }, [auctionId, amount])
+
+    const PlaceBidCallback = useCallback(async () => {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Foundation/hooks/usePlaceBidCallback.ts:35]:
```suggestion
    const placeBidCallback = useCallback(async () => {
```

**Comment by AinuX** [comment]
@lucasespinosa28

**Comment by lucasespinosa28** [comment]
> @lucasespinosa28
> 
> Could you share info how to access the testnet of Foundation (https://goerli.foundation.app/)? I have no idea where to apply the visitor password? thanks !

hi @AinuX, how to test foundation in goerli

1. access the foundation NFT contract in: https://goerli.etherscan.io/address/0x44458837ac4036337e5ce46ce28a744e05e02016#writeProxyContract
2. mint a new NFT and approve to use in the market, use this metadata already store in foundation IPFS, QmaT5krJGYkj6pkBt5GuMRDzZoFF2oThd9JZ9yxXwi4hec/metadata.js
Captura de tela 2021-11-12 170644 (2)
3. now add the NFT you minted to auction in the market,https://goerli.etherscan.io/address/0xeb1bd095061bbdb1ad065524628812cae63e4222#writeProxyContract
Captura de tela 2021-11-12 171147 (2)
4. in Twitter writer this, https://gorli.foundation.app/@testing/test- {id of NFT did you put to auction}
Captura de tela 2021-11-12 171352 (2)

**Review by lucasespinosa28 — COMMENTED** [review]
added format address and test in gorli

**Comment by wenluomask** [comment]
Thanks for your quick reply and detailed instruction! Could you please rebuild your pr since the latest build is failed?

**Review by lucasespinosa28 — COMMENTED** [review]
@wenluomask , the build error was fixed

**Comment by AinuX** [comment]
@lucasespinosa28 
Found some issues here, please take a look, thanks!

1. Style issue
image

2. Fail to display if the NFT is stream media(like mp4, etc.)
image

**Comment by AinuX** [comment]
@lucasespinosa28 
Since the current UI implementation is a bit different from our Collectibles plugin,  please refer to our figma to do some adjustments, thanks!
figma link: https://www.figma.com/file/gVkQ67y285b4FXVV1KPThN/TwitterV1?node-id=8896%3A0

**Review by lucasespinosa28 — COMMENTED** [review]
@AinuX, I recreate the UI using the refer from Figma, I did some changes  because of the limitation of foundation API and material UI 

the problem of not showing mp4 NFT only happen in testnet because the foundation does not have a video thumbnail for test NFT, in mainnet doesn't have this problem

https://twitter.com/search?q=%20%20%40withFND&src=typed_query&f=top
use this link to test the UI in mainnet
use the testnet version to test buy the NFT

https://user-images.githubusercontent.com/52639395/142742772-fd05a984-461b-434d-8f8e-4dce1d8ce9fb.mp4

**Review by lucasespinosa28 — COMMENTED** [review]
resolved conflict

**Review by lucasespinosa28 — COMMENTED** [review]
I fix the conflicts

**Review by lucasespinosa28 — COMMENTED** [review]
fix spell and media query

**Review by lucasespinosa28 — COMMENTED** [review]
fix conflict

**Review by lucasespinosa28 — COMMENTED** [review]
conflict fixed

**Review by lucasespinosa28 — COMMENTED** [review]
fix parseURL path

**Review by lucasespinosa28 — COMMENTED** [review]
conflict fixed

**Review by lucasespinosa28 — COMMENTED** [review]
conflict fixed

**Review by lucasespinosa28 — COMMENTED** [review]
fixed register plugins

**Review by lucasespinosa28 — COMMENTED** [review]
fixed problem with foundation

**Review by lucasespinosa28 — COMMENTED** [review]
fixed eslint errors

**Review by lucasespinosa28 — COMMENTED** [review]
eslint errors fixed

**Comment by LindaMask** [comment]
@lucasespinosa28 hey are you on telegram or discord? can you reach out to me? Telegram:@lin_dao Discord:lin_DAO_#6015

File: packages/mask/src/plugins/Foundation/utils.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+        return {
+            chainId: _url.host.includes('gorli') ? ChainId.Gorli : ChainId.Mainnet,
+            link: url,
+        }
+    }
+    return
+}
+
+export function convertDate(unix: string) {
+    const date = new Date(Number.parseInt(unix, 10) * 1000)
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Foundation/utils.ts:31]: Recommend to use `date-fns`.

File: packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationPlaceBid.tsx [hunk 18]
```diff
@@ -60,33 +60,33 @@ function HasCountdown(props: Countdown) {
 }
 
 function FoundationPlaceBid(props: Props) {
-    //#region context
+    // #region context
     const { t } = useI18N()
     const account = useAccount()
     const chainId = useChainId()
-    const testNet = chainId === 5 ? true : false
+    const testNet = chainId === 5
```
Comment (septs) [inline packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationPlaceBid.tsx:67]:
```suggestion
    const testNet = chainId === ChainId.Gorli
```

File: packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationImage.tsx [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+    if (isVideo) {
+        return (
+            <img
+                className={classes.image}
+                src={`https://assets.foundation.app/${path[0]}/${path[1]}/${path[2]}/nft.jpg`}
+                alt={alt}
+            />
+        )
+    }
+    const is3D = src?.match(/\.(glb|gltf)$/i)
```
Comment (septs) [inline packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationImage.tsx:54]:
```suggestion
    const isModel = src?.match(/\.gl(b|tf)$/i)
```

File: packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationImage.tsx [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+            />
+        )
+    }
+    return <img className={classes.image} src={src} alt={alt} />
+}
+
+function FoundationImage(props: Props) {
+    const { classes } = useStyles()
+    const image = props.metadata.image.split('/')
+    const src = `${props.nftContract.baseURI}${image.at(-2)}/${image.at(-1)}`
```
Comment (septs) [inline packages/mask/src/plugins/Foundation/SNSAdaptor/FoundationImage.tsx:70]: plz use https://npm.im/urlcat build the url

**Pull request closed without merging**